kong/tools/stream_api.lua

```diff
@@ -0,0 +1,77 @@
+
+require "lua_pack"
+
+
+local kong       = kong
+local st_pack    = string.pack      -- luacheck: ignore string
+local st_unpack  = string.unpack    -- luacheck: ignore string
+local st_format  = string.format
+
+local PREFIX = ngx.config.prefix()
+
+local stream_api = {}
+
+local _handlers  = {}
+
+
+function stream_api.register(k, f)
+  _handlers[k] = f
+end
+
+function stream_api.request(key, data, socket_path)
+  local socket = assert(ngx.socket.udp())
+  assert(socket:setpeername(socket_path or "unix:" .. PREFIX .. "/stream_rpc.sock"))
+
```

Comment:
We should check for the `type` of the arguments here and throw an error if they mismatch (to aid debugging).

Also, maybe a size check on the argument is a good idea. I Googled around and it seems that there is a limit of datagram size that could be sent over domain sockets, however the number seems to vary depends on Kernel params. https://stackoverflow.com/questions/21856517/whats-the-practical-limit-on-the-size-of-single-packet-transmitted-over-domain

Maybe something a bit conservative like 128KB at first, we can always increase this in the future if needed.

Reply:
the typical limit i've seen is like 64KiB, but somewhere in openresty docs mentions some buffer limit of 8KiB.  I guess it's for UDP and not unix sockets but i haven't checked the sources.  added a test for 8000, which i guess should be more than enough for requests.

big responses are slightly more worrying, but i guess hard failing wouldn't help in debugging.  I've put just a WARN log, hope it's enough